### ISSUE TITLE
rerun-if-env-changed=CARGO_MANIFEST_DIR doesn't trigger a rebuild on change

### ISSUE THREAD

**Issue by alecmocatta** [rust-lang/rust#76511]
`CARGO_MANIFEST_DIR` changing doesn't trigger a rebuild, even if `rerun-if-env-changed=CARGO_MANIFEST_DIR` is specified in the `build.rs`.

Thus `env!("CARGO_MANIFEST_DIR")` and `env::var("CARGO_MANIFEST_DIR").unwrap()` can differ in binaries executed by `cargo run/test/bench`.

**Comment by RalfJung** [comment]
@alecmocatta which version of rustc and cargo did you try this with?
I thought these bugs would be fixed by https://github.com/rust-lang/rust/pull/71858, but maybe that has not reached stable yet.

**Comment by alecmocatta** [comment]
Latest nightly, 2020-09-09

**Comment by RalfJung** [comment]
Cc @petrochenkov @ehuss

**Comment by petrochenkov** [comment]
This is an issue (or non-issue) with cargo rather than rustc.
Environment variables *known* to cargo are treated differently than user-defined ones, for example https://github.com/rust-lang/cargo/pull/8421 seems to filter away at least some of them from dependencies and not rebuild on their change.

**Comment by alecmocatta** [comment]
Yes sorry this is a potential issue with cargo, I was too hasty in my filing here. Is it possible to move the issue?

Alternately as this is a known issue as @RalfJung tracked down https://github.com/rust-lang/cargo/pull/8421/files#r485577035 feel free to close if appropriate.

**Comment by RalfJung** [comment]
It is possible to move this to the cargo repo but I do not have the required permissions.

IMO this issue reflects a reasonable request, so I'd keep it open for now. I'd probably change the title; I would not even expect `rerun-if-env-changed=CARGO_MANIFEST_DIR` to be needed.
